vim-patch:8.2.{3430,3434,3462,3463,3555,3609,3610}: ModeChanged autocmd

vimpostor opened the pull request:
This is a direct port of the generic `ModeChanged` autocmd, that I already implemented for vim.
The autocmd can pattern match against `old_mode:new_mode`, for example:

- To only trigger, when switching from visual mode to visual-line mode, use `autocmd ModeChanged v:V`
- To only trigger, when entering normal mode from any other mode, use `autocmd ModeChanged *:n`

This patch is best tested using:
```bash
build/bin/nvim --clean -Nu <(cat<<EOF
set noshowmode
au ModeChanged * echomsg printf('the old mode was: "%s", the new mode is: "%s"', v:event.old_mode, v:event.new_mode)
EOF
)
```

Let me know if you need any changes! :)

vimpostor committed:
vim-patch:8.2.3430: no generic way to trigger an autocommand on mode change

Problem:    No generic way to trigger an autocommand on mode change.
Solution:   Add the ModeChanged autocommand event. (Magnus Gross, closes vim/vim#8856)
https://github.com/vim/vim/commit/f1e8876fa2359b572d262772747405d3616db670

N/A patches for version.c:

vim-patch:8.2.3434: function prototype for trigger_modechanged() is incomplete

Problem:    Function prototype for trigger_modechanged() is incomplete.
Solution:   Add "void".
https://github.com/vim/vim/commit/28e591dd5080bbcd0f468f9d9597cedb716e28c9

Fixes #4399.
Fixes #7416.

vimpostor commented:
A little note for the curious reader: This autocmd is obviously intended to be triggered for **every** user-visible modechange. Unfortunately due to some previous weird design decisions, we did not yet come up with a sane way to map the Terminal-Normal  <-> Terminal-Job switch into this autocmd. You can read more about this [here](https://github.com/vim/vim/pull/8856#issuecomment-917718219) and feel free to chime in over there (or here), if you know a good way to solve that problem.

clason commented:
(Neovim terminal and vim terminal are very different beasts, anyway.)

zeertzjq commented on src/nvim/misc1.c:
```diff
@@ -1067,3 +1067,32 @@ void add_time(char_u *buf, size_t buflen, time_t tt)
+
+  dict_T *v_event = get_vim_var_dict(VV_EVENT);
+
+  char *mode = get_mode();
```
In your original vim patch you passed `VAR_UNKNOWN` to `f_mode` so it only returns one character, but here you get the full mode string. Is one of them wrong?

Also, `get_mode` is in `state.c`, maybe you should put this function in `state.c` as well.

zeertzjq replied:
> In your original Vim patch you passed `VAR_UNKNOWN` to `f_mode` so it only returns one character, but here you get the full mode string. Is one of them wrong?
> 
> Also, `get_mode` is in `state.c`, maybe you should put this function in `state.c` as well.

I guess your original Vim patch was wrong, as you allocated `MODE_MAX_LENGTH` bytes for `last_mode`.

zeertzjq replied:
> In your original Vim patch you passed `VAR_UNKNOWN` to `f_mode` so it only returns one character, but here you get the full mode string. Is one of them wrong?

I guess you passed wrong arguments to `f_mode` in your original Vim patch, as you allocated `MODE_MAX_LENGTH` bytes for `last_mode`.

zeertzjq commented on runtime/doc/autocmd.txt:
```diff
@@ -729,7 +729,22 @@ MenuPopup			Just before showing the popup menu (under the
+				The following values of |v:event| are set:
+					old_mode The mode before it changed.
+					new_mode The new mode as also returned
+					         by |mode()|.
```
In case you found out you original Vim patch was wrong, this should `mode(1)` instead of `mode()`.

zeertzjq commented on runtime/doc/autocmd.txt:
```diff
@@ -729,7 +729,22 @@ MenuPopup			Just before showing the popup menu (under the
+				last triggered.
+				Usage example to use relative line numbers
+				when entering visual mode: >
+					:autocmd ModeChanged *:v set rnu
```
In case you find out your original Vim patch was wrong, this should be `*:v*` instead of `*:v`.

zeertzjq commented on src/nvim/misc1.c:
```diff
@@ -1067,3 +1067,32 @@ void add_time(char_u *buf, size_t buflen, time_t tt)
 }
+
+/// Fires a ModeChanged autocmd.
+void trigger_modechanged(void)
```
`get_mode` is in `state.c`, so maybe you can put this in `state.c` as well.

janlazo replied:
Moving can be done in a separate PR after the patches for moving code to separate files are done to ease porting.

vimpostor commented:
> In your original Vim patch you passed VAR_UNKNOWN to f_mode so it only returns the first character of the mode string, but here you get the full mode string. Is one of them wrong?

This is a good catch. I actually added `VAR_UNKNOWN` to make `vim's` vim9 check not return:
```c
    if (in_vim9script() && check_for_opt_bool_arg(argvars, 0) == FAIL)
        return;
```

I didn't notice that I fixed the mode length to 1 with this as well. Right now, there is no `trigger_modechanged()` that would yield a minor mode, but I think it makes sense to at least be able to do this.

So I guess I will make a vim patch that fixes that problem, then wait for it to be upstream, and then include it as part of this PR.

vimpostor commented:
Or is it perhaps better to only make major mode changes count?

zeertzjq commented:
> Right now, there is no `trigger_modechanged()` that would yield a minor mode

I compiled Neovim on your branch, and found out that `ModeChanged` yields `niI` as `new_mode` when pressing `CTRL-O` in Insert mode.

zeertzjq commented:
A suggestion: maybe you split out a `get_mode` function from `f_mode` in Vim source code, just like Neovim's `get_mode` and `f_mode` functions.

clason commented:
You probably want to add https://github.com/vim/vim/commit/72406a4bd2896915b6f541e26d41521a59b1f846?

zeertzjq commented:
There is also <https://github.com/vim/vim/commit/d85931e67316e824878000d0ead122553ccef3a4>.

vimpostor commented:
Correct. I will include these two patches tomorrow.

zeertzjq commented:
For terminal normal mode maybe you need to check `curbuf->terminal`.

clason commented:
_and_ https://github.com/vim/vim/commit/72d2fa69e56dd0e61eac87a9c3b79cd89d47cdd9 :P

zeertzjq commented:
Patch 8.2.3461 is unrelated to `ModeChanged`, and this PR is gonna take some time, so I'm porting it in another PR #15878.

zeertzjq commented:
Also, maybe in Neovim it is easy to detect Terminal-Normal <-> Terminal-Job mode change? I haven't tried yet.

justinmk commented:
> Also, maybe in Neovim it is easy to detect Terminal-Normal <-> Terminal-Job mode change? I haven't tried yet.

There is no "terminal-normal", it's just normal-mode. "Terminal-Job" mode is called `:help terminal-mode`, basically "insert-mode in a terminal buffer".

zeertzjq commented:
> > Also, maybe in Neovim it is easy to detect Terminal-Normal <-> Terminal-Job mode change? I haven't tried yet.
> 
>     * There is no "terminal-normal", it's just normal-mode. Normal-mode in a `:terminal` buffer is not special, in Nvim.
> 
>     * "Terminal-Job" mode is called `:help terminal-mode`, basically "insert-mode in a terminal buffer".

But in a `:terminal`, in normal mode pressing `i` goes to Terminal-mode instead of Insert mode, so it is indeed different from other Normal mode.

vimpostor committed:
vim-patch:8.2.3462: ModeChanged only uses one character for new_mode and old_mode

Problem:    The ModeChanged event only uses one character for the new_mode and
            old_mode values.
Solution:   Pass one as first argument to mode(). (issue vim/vim#8856)
https://github.com/vim/vim/commit/d85931e67316e824878000d0ead122553ccef3a4

vimpostor committed:
vim-patch:8.2.3463: pattern matching with ModeChanged not tested

Problem:    Pattern matching with ModeChanged not tested.
Solution:   Add a few more test lines. (issue vim/vim#8856)
https://github.com/vim/vim/commit/72d2fa69e56dd0e61eac87a9c3b79cd89d47cdd9

zeertzjq commented:
Note to reviewers: another Vim patch is needed to trigger `ModeChanged` on minor mode change.

vimpostor committed:
vim-patch:8.2.3555: ModeChanged is not triggered on every mode change

Problem:    ModeChanged is not triggered on every mode change.
Solution:   Also trigger on minor mode changes. (Maguns Gross, closes vim/vim#8999)
https://github.com/vim/vim/commit/25def2c8b8bd7b0c3d5f020207c717a880b05d50

vimpostor committed:
feat: trigger ModeChanged for terminal modes

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -906,6 +906,7 @@ static uint8_t *command_line_enter(int firstc, long count, int indent)
     }
     tl_ret = true;
   }
+  trigger_modechanged();
```
Patch v8.2.3430 had

```c
if (!debug_mode)
  trigger_modechanged();
```

I guess that `debug_mode` should not be in the vim patch. It's rarely used throughout the codebase.

vimpostor replied:
The `debug_mode` is only there, because some weird behaviour in some test caused the test to fail without it. See https://github.com/vim/vim/pull/8856#issuecomment-916805623 for reference. This is not the case for neovim, so I exclude `debug_mode` in the port.

janlazo commented on src/nvim/globals.h:
```diff
@@ -727,6 +727,7 @@ EXTERN bool listcmd_busy INIT(= false);     // set when :argdo, :windo or
                                             // :bufdo is executing
 EXTERN bool need_start_insertmode INIT(= false);
 // start insert mode soon
+EXTERN char *last_mode INIT(= NULL);
```
Diverges from the patch.
I don't mind because of the NULL check in trigger_modechanged().

vimpostor replied:
This diverges because I am using an array in vim and a pointer in the port here. I use a pointer here, because `get_mode()` returns allocated memory. I agree, that this is fine because of the NULL check.

vimpostor replied:
BTW the null check is removed again in the port of 8.2.3555, because `last_mode` cannot ever be `NULL` anymore, as it is now initialized in `autocmd.c` when setting up the first `ModeChanged` autocmd.

janlazo commented:
Port the following for `get_mode()`. 4 is hardcoded.

```c
#define MODE_MAX_LENGTH	4	// max mode length returned in mode()
```

janlazo commented on src/nvim/autocmd.c:
```diff
@@ -925,6 +925,18 @@ static int do_autocmd_event(event_T event, char_u *pat, bool once, int nested, c
+          } else {
+            STRCPY(last_mode, mode);
+          }
+          xfree(mode);
```
`get_mode()` is an allocated string.
What's the issue with freeing `last_mode` first and then assigning the new value to it?

zeertzjq replied:
In the original Vim patch `last_mode` is an array. If `last_mode` is changed to an array here then this will make sense.

vimpostor replied:
See https://github.com/neovim/neovim/pull/15840#discussion_r740912878 for why I use a pointer instead of an array.

I agree with @janlazo that just freeing and assigning is better. The if branch comes from a misunderstanding of mine, which was that I thought freeing NULL would be a problem (which it obviously isn't).
I have implemented the proposed change now.

janlazo reviewed (changes requested):
It's a good port.
Address my comments, and I'll review and maybe merge.

vimpostor commented:
> Port the following for `get_mode()`. 4 is hardcoded.
> 
> ```c
> #define MODE_MAX_LENGTH	4	// max mode length returned in mode()
> ```

Implemented in the port of 8.2.3430.

> It's a good port. Address my comments, and I'll review and maybe merge.

All comments have been addressed now, thanks for reviewing! ;)

zeertzjq commented on src/nvim/misc1.c:
```diff
@@ -1059,3 +1059,33 @@ void add_time(char_u *buf, size_t buflen, time_t tt)
+  xfree(pat_pre);
+
+  apply_autocmds(EVENT_MODECHANGED, pat, NULL, false, curbuf);
+  STRCPY(last_mode, mode);
+
+  xfree(pat);
+  tv_dict_free_contents(v_event);
+  hash_init(&v_event->dv_hashtab);
+  xfree(mode);
```
Hmm, I think this is what janlazo wanted you to do:
```suggestion
  xfree(last_mode);
  last_mode = mode;

  xfree(pat);
  tv_dict_free_contents(v_event);
  hash_init(&v_event->dv_hashtab);
```

vimpostor replied:
Thanks, I only did the equivalent change in `autocmd.c` and totally forgot about the main part. This is done now and squashed into the port of 8.2.3430 (the first commit in this series)

vimpostor commented:
@gpanders Do you mind retriggering the CI?

vimpostor commented:
I do not know why the CI sometimes fails in the Windows build matrix and I do not have any Windows system available anywhere to reproduce this locally.
But I guess since the current master HEAD has a failing Windows CI too, this is not a problem introduced by my patch, correct?

zeertzjq commented:
It began happening since #16041 was merged.

clason commented:
But that's just correlation, not causation! (MSVC has been flaky before, and the MingW failures were fixed in #16187 -- you should rebase your PR on master to pick up the fix.)

vimpostor commented:
> But that's just correlation, not causation! (MSVC has been flaky before, and the MingW failures were fixed in #16187 -- you should rebase your PR on master to pick up the fix.)

I already rebased on that, that commit was included in the last CI run.
I have rebased now again on master nonetheless, but I don't think any of the new 3 commits 2481b18a5..2481b18a5~2 have anything to do with this.

vimpostor commented:
For reference, the last Windows CI failed with:
```
2021-11-03T00:07:03.5618809Z   [ RUN      ] API nvim_parse_expression works with + and @a: -- Tests exited non-zero: Access violation
2021-11-03T00:07:03.5622656Z   CMake Error at D:/a/neovim/neovim/cmake/RunTests.cmake:94 (message):
2021-11-03T00:07:03.5624978Z CUSTOMBUILD : functional tests failed with error : Access violation [D:\a\neovim\neovim\build\functionaltest.vcxproj]
```

I believe this may just be a flaky test. I can't reproduce on Linux with `TEST_FILE=test/functional/api/vim_spec.lua make functionaltest` and I have no Windows system available so can someone restart the CI?

vimpostor commented:
Thanks for retriggering.
Looks like it was indeed just a flaky Windows test, and the two other CI failures (MINGW_32, MSVC_32) match up with the current master branch, so are not introduced by this PR.

@janlazo Highly appreciate it if you could take a look at the changes that I did in regards to your comments :)

zeertzjq commented:
I just noticed you forgot the changes in `apply_autocmds_group`.

zeertzjq commented:
> I just noticed you forgot the changes in `apply_autocmds_group`.

@vimpostor

vimpostor commented:
> I just noticed you forgot the changes in `apply_autocmds_group`.

Yes, somehow the `vim-patch.sh` script was missing that hunk. I have added it now and squashed it into the first commit of this series.

Exact diff that has been squashed:
```diff
diff --git a/src/nvim/autocmd.c b/src/nvim/autocmd.c
index 24f4cbe7e..490fe5a0a 100644
--- a/src/nvim/autocmd.c
+++ b/src/nvim/autocmd.c
@@ -1447,7 +1447,7 @@ static bool apply_autocmds_group(event_T event, char_u *fname, char_u *fname_io,
   // invalid.
   if (fname_io == NULL) {
     if (event == EVENT_COLORSCHEME || event == EVENT_COLORSCHEMEPRE
-        || event == EVENT_OPTIONSET) {
+        || event == EVENT_OPTIONSET || event == EVENT_MODECHANGED) {
       autocmd_fname = NULL;
     } else if (fname != NULL && !ends_excmd(*fname)) {
       autocmd_fname = fname;
@@ -1501,11 +1501,12 @@ static bool apply_autocmds_group(event_T event, char_u *fname, char_u *fname_io,
         || event == EVENT_CMDWINLEAVE || event == EVENT_CMDUNDEFINED
         || event == EVENT_COLORSCHEME || event == EVENT_COLORSCHEMEPRE
         || event == EVENT_DIRCHANGED || event == EVENT_FILETYPE
-        || event == EVENT_FUNCUNDEFINED || event == EVENT_OPTIONSET
-        || event == EVENT_QUICKFIXCMDPOST || event == EVENT_QUICKFIXCMDPRE
-        || event == EVENT_REMOTEREPLY || event == EVENT_SPELLFILEMISSING
-        || event == EVENT_SYNTAX || event == EVENT_SIGNAL
-        || event == EVENT_TABCLOSED || event == EVENT_WINCLOSED) {
+        || event == EVENT_FUNCUNDEFINED || event == EVENT_MODECHANGED
+        || event == EVENT_OPTIONSET || event == EVENT_QUICKFIXCMDPOST
+        || event == EVENT_QUICKFIXCMDPRE || event == EVENT_REMOTEREPLY
+        || event == EVENT_SPELLFILEMISSING || event == EVENT_SYNTAX
+        || event == EVENT_SIGNAL || event == EVENT_TABCLOSED
+        || event == EVENT_WINCLOSED) {
       fname = vim_strsave(fname);
     } else {
       fname = (char_u *)FullName_save((char *)fname, false);
```

clason commented:
Also related: https://github.com/vim/vim/commit/3075a45592fe76f2febb6321632a23e352efe949

vimpostor committed:
vim-patch:8.2.3609: internal error when ModeChanged is triggered recursively

Problem:    Internal error when ModeChanged is triggered when v:event is
            already in use.
Solution:   Save and restore v:event if needed.
https://github.com/vim/vim/commit/3075a45592fe76f2febb6321632a23e352efe949

In the vim codebase there is no occurrence of get_vim_var_dict(VV_EVENT)
after the above patch, so in order to hold the same invariant in the
neovim codebase we needed to replace more occurrences than the related
vim patch.

vimpostor commented:
And https://github.com/vim/vim/commit/a062006b9de0b2947ab5fb376c6e67ef92a8cd69

In which file should I put `save_v_event_T` from patch 8.2.3609? `structs.h` does not exist. `src/nvim/event/defs.h` is by the name the best match of all `defs.h`, but it looks like it is for other events intended.

vimpostor committed:
vim-patch:8.2.3610: crash when ModeChanged triggered too early

Problem:    Crash when ModeChanged triggered too early.
Solution:   Trigger ModeChanged after setting VIsual.
https://github.com/vim/vim/commit/a062006b9de0b2947ab5fb376c6e67ef92a8cd69

vimpostor commented:
Re-requesting review because two patches have been added (two last commits).

zeertzjq commented:
`src/nvim/event/` should be for the event-loop, which has nothing to do with autocommands. `src/nvim/buffer_defs.h` may be more appropriate. (as there are also things like `ChangedtickDictItem` there)

zeertzjq commented:
More occurrences of `get_vim_var_dict(VV_EVENT)` may need to be replaced by `get_v_event(&save_v_event)`.

vimpostor commented:
- Moved the `save_v_event_T` struct to `src/nvim/buffer_defs.h`
- Replaced more occurrences of `get_vim_var_dict(VV_EVENT)`. After the vim patch there are no more occurrences of `get_vim_var_dict(VV_EVENT)`, so the same invariant holds now in the neovim codebase (squashed into 8.2.3609)

zeertzjq commented on src/nvim/buffer_defs.h:
```diff
@@ -1470,4 +1470,10 @@ static inline int win_hl_attr(win_T *wp, int hlf)
+  bool sve_did_save;
+  hashtab_T sve_hashtab;
+} save_v_event_T;
+
```
This probably shouldn't be put at the bottom of this file, especially considering the `Macros defined in Vim, but not in Neovim` comment above.

vimpostor replied:
Moved it up a bit, I hope the new place is good for you.

vimpostor commented:
@janlazo Is there anything else that needs to be done here?

vimpostor commented:
@janlazo I don't think the CI is failing because of my patch, as the linting errors are in completely different files, but if you want I can rebase on latest master.

vimpostor commented:
@janlazo Sorry for asking again, but any chance of getting this in before the upcoming 0.6 release?

clason commented:
Probably not; that is a bit too large to merge three days before the release...

janlazo commented on src/nvim/misc1.c:
```diff
@@ -1059,3 +1059,58 @@ void add_time(char_u *buf, size_t buflen, time_t tt)
+  }
+
+  char *mode = get_mode();
+  if (STRCMP(mode, last_mode) == 0) {
```
This assumes that `last_mode` is never NULL because it is set elsewhere before `has_event(EVENT_MODECHANGED)`  evaluates to true. Could be a false positive in static analyzer.

janlazo replied:
Could have used strncmp because of `MODE_MAX_LENGTH`.
